mage_ai/streaming/sources/nats_js.py

```diff
@@ -0,0 +1,187 @@
+                "closed_cb": self.closed_cb,
+            }
+
+            # Configure SSL context if use_tls is True
+            if self.config.use_tls and self.config.ssl_config:
+                ssl_ctx = ssl.create_default_context(purpose=ssl.Purpose.SERVER_AUTH)
+                if self.config.ssl_config.cafile:
+                    ssl_ctx.load_verify_locations(self.config.ssl_config.cafile)
+                if self.config.ssl_config.certfile and self.config.ssl_config.keyfile:
+                    ssl_ctx.load_cert_chain(
+                        certfile=self.config.ssl_config.certfile,
+                        keyfile=self.config.ssl_config.keyfile
+                    )
+                connect_opts["tls"] = ssl_ctx
+
+            # Use NKEY if provided
+            if self.config.nkeys_seed_str:
+                connect_opts["nkeys_seed_str"] = self.config.nkeys_seed_str
+
+            # Establish connection with the configured options
+            self.nc = await nats.connect(**connect_opts)
+            self.js = self.nc.jetstream()
+
+            # Default consumer_name to stream_name if not provided
+            consumer_name = self.config.consumer_name or self.config.stream_name
+
+            await self.js.add_stream(name=self.config.stream_name, subjects=[self.config.subject])
+            self.psub = await self.js.pull_subscribe(self.config.subject, consumer_name)
+        except NoServersError as e:
+            self._print(f'Caught NoServersError while connecting to NATS server: {e}')
+        except Exception as e:
+            self._print(f'Caught exception while connecting to NATS server: {e}')
+
+    # Define callback methods
+    async def disconnected_cb(self):
+        self._print('Got disconnected!')
+
+    async def reconnected_cb(self):
+        self._print(f'Got reconnected to {self.nc.connected_url.netloc}')
+
+    async def error_cb(self, e):
+        self._print(f'There was an error: {e}')
+
+    async def closed_cb(self):
+        self._print('Connection is closed')
+
+    async def aclose_client(self):
+        await self.nc.close()
+
+    def init_client(self):
+        future = asyncio.run_coroutine_threadsafe(self.ainit_client(), self.loop)
+        future.result()
+
+    def close_client(self):
+        asyncio.run_coroutine_threadsafe(self.aclose_client(), self.loop)
+
+    def batch_read(self, handler: Callable):
+        self.init_client()
+
+        try:
+            while True:
+                self._print("Fetching messages...")
+                message_tuples = self.fetch_messages()
+                self._print(f"Fetched {len(message_tuples)} messages")
+
+                if not message_tuples:
+                    self._print("No messages fetched, continuing to next iteration")
+                    continue
+
+                # Accumulate all messages in a batch
+                batch = [json.loads(msg.data.decode()) for _, msg in message_tuples]
```

Comment:
You already decode the data in `afetch_messages`. Why do you decode it again?

Reply:
addressed with 94d93afbd